src/main.c

```diff
@@ -80,6 +100,129 @@ make_conn(const char *hostname)
   return -1;
 }
 
+/*
+ * Copy all the ciphertext bytes from buf into the client session.
+ * Returns 0 for success, 1 for error.
+ */
+int
+copy_tls_bytes_into_client_session(
+  struct rustls_client_session *client_session, uint8_t *buf, size_t len)
+{
+  ssize_t n;
+  int result;
+
+  while(len > 0) {
+    n = rustls_client_session_read_tls(client_session, (uint8_t *)buf, len);
+    if(n == 0) {
+      fprintf(stderr, "EOF from ClientSession::read_tls %ld \n", len);
+      return 1;
+    }
+    else if(n < 0) {
+      fprintf(stderr, "Error in ClientSession::read_tls\n");
+      return 1;
+    }
+    if((size_t)n > len) {
+      fprintf(stderr, "too many bytes written to ClientSession; overflow\n");
+      return 1;
+    }
+    len -= n;
+    buf += n;
+
+    result = rustls_client_session_process_new_packets(client_session);
+    if(result != RUSTLS_RESULT_OK) {
+      fprintf(stderr, "Error in process_new_packets\n");
+      return 1;
+    }
+  }
+  return 0;
+}
+
+/* Read all available bytes from the client_session until EOF.
+ * Note that EOF here indicates "no more bytes until
+ * process_new_packets", not "stream is closed".
+ *
+ * Returns 0 for success, 1 for error.
+ */
+int
+copy_plaintext_to_stdout(struct rustls_client_session *client_session)
+{
+  int result = 1;
+  char buf[2048];
+  int n;
+
+  for(;;) {
+    bzero(buf, sizeof(buf));
+    n =
+      rustls_client_session_read(client_session, (uint8_t *)buf, sizeof(buf));
+    if(n == 0) {
+      /* EOF from ClientSession::read. This is expected. */
+      return 0;
+    }
+    else if(n < 0) {
+      fprintf(stderr, "Error in ClientSession::read\n");
+      return 1;
+    }
+
+    result = write_all(STDOUT_FILENO, buf, n);
+    if(result != 0) {
+      return result;
+    }
+  }
+
+  return 0;
+}
+
+/*
+ * Do one read from the socket, and process all resulting bytes into the
+ * client_session, then copy all plaintext bytes from the session to stdout.
+ * Returns CRUSTLS_DEMO_OK for success, CRUSTLS_DEMO_ERROR for error, and
+ * CRUSTLS_DEMO_AGAIN if we go an EAGAIN or EWOULDBLOCK reading from the
+ * socket.
+ */
+enum crustls_demo_result
+do_read(int sockfd, struct rustls_client_session *client_session)
+{
+  int result = 1;
+  ssize_t n = 0;
+  char buf[2048];
+
+  bzero(buf, sizeof(buf));
+  n = read(sockfd, buf, sizeof(buf));
+  if(n == 0) {
+    fprintf(stderr, "EOF reading from socket\n");
+    return CRUSTLS_DEMO_ERROR;
+  }
+  else if(n < 0) {
+    if(errno == EAGAIN || errno == EWOULDBLOCK) {
+      fprintf(stderr, "reading from socket: EAGAIN or EWOULDBLOCK\n");
```

Comment:
nit: consider `strerror(3)` to get a natural language description of `errno`. Then again it can often be nice to have to specific errno. Otherwise it's hard to tell `EACCES` from `EPERM`: "Permission denied" and "Operation not permitted", respectively.